### PR TITLE
RELATED: SD-781 - Too many data points

### PR DESCRIPTION
**Pull request by ptquang86**
---

[Check PR owner responsibilities](https://confluence.intgdc.com/display/Development/Code-reviews#Code-reviews-Ownerresponsibilities)

Supported PR commands:

Command | Description
--- | ---
`extended test - examples` | Live examples tests
`extended test - storybook` | Storybook screenshot tests

See more [options](https://confluence.intgdc.com/display/kbhr/How+to+work+with+git+and+Github#HowtoworkwithgitandGithub-Extendedchecks).

# PR Checklist

- [x] Verify code changes ([Checklist](https://confluence.intgdc.com/display/Development/Code-reviews+checklist), [Best practices](https://confluence.intgdc.com/display/Development/Code-reviews+best+practices))
- [x] [Verify pull-request formalities](https://confluence.intgdc.com/display/Development/Code-reviews)
- [ ] Change was tested by using [gdc-dev-release](https://confluence.intgdc.com/display/~tomas.vojtasek/Private+NPM) in [gdc-analytical-designer](https://github.com/gooddata/gdc-analytical-designer) and [gdc-dashboards](https://github.com/gooddata/gdc-dashboards) (if applicable)
- [x] Migration guide (for major changes) is mentioned in [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [x] Successful `extended test - examples`
- [x] Successful `extended test - storybook`
- [x] Checked yarn.lock consistency (no dep. duplicities especially Goodstrap)

# Related PRs

- gdc-analytical-designer:
- gdc-dashboards:

# Related Jira tasks
- SD-781: https://jira.intgdc.com/browse/SD-781

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5674/ : SUCCESS in 9m 08s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1214/ : SUCCESS in 7s

**Comment by ptquang86** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1324.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1215/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2323/ : SUCCESS in 21m 39s

**Comment by ptquang86** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2902/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1216/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2324/ : SUCCESS in 13m 52s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5677/ : SUCCESS in 9m 52s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1220/ : SUCCESS in 5s

File: src/components/core/GeoChart.tsx [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+ */
+export class GeoChartInner extends BaseVisualization<IGeoChartInnerProps, {}> {
+    public static defaultProps: Partial<IGeoChartInnerProps> = {
+        ...commonDefaultProps,
+        chartRenderer: renderChart,
+        legendRenderer: renderLegend,
+    };
+
+    public componentDidUpdate(prevProps: IGeoChartInnerProps) {
+        if (this.props.execution !== prevProps.execution) {
```
Comment (dlcd2203) [inline src/components/core/GeoChart.tsx]: "this.props.execution !== prevProps.execution" => can be wrong.

File: src/components/core/geoChart/GeoChartRenderer.tsx [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    public static defaultProps: Partial<IGeoChartRendererProps> = {
+        config: {},
+    };
+
+    private chart: mapboxgl.Map;
+    private chartRef: HTMLElement;
+
+    public componentDidUpdate(prevProps: IGeoChartRendererProps) {
+        const { execution } = this.props;
+        if (execution !== prevProps.execution) {
```
Comment (dlcd2203) [inline src/components/core/geoChart/GeoChartRenderer.tsx]: I think the result is incorrect when using === operator to compare 2 objects

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5684/ : SUCCESS in 5m 11s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1230/ : SUCCESS in 5s

File: src/components/core/GeoChart.tsx (hunk 1 above)
Comment (ptquang86) [reply src/components/core/GeoChart.tsx]: fixed

File: src/components/core/geoChart/GeoChartRenderer.tsx (hunk 2 above)
Comment (ptquang86) [reply src/components/core/geoChart/GeoChartRenderer.tsx]: fixed

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5685/ : SUCCESS in 4m 58s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1231/ : SUCCESS in 4s

File: src/components/core/GeoChart.tsx [hunk 3]
```diff
@@ -44,6 +48,16 @@ export class GeoChartInner extends BaseVisualization<IGeoChartInnerProps, {}> {
         legendRenderer: renderLegend,
     };
 
+    public componentDidUpdate(prevProps: IGeoChartInnerProps) {
+        if (!isEqual(this.props.execution, prevProps.execution)) {
```
Comment (naibinh) [inline src/components/core/GeoChart.tsx]: I think we should put this checking to `shouldComponentUpdate`

File: src/helpers/geoChart.ts [hunk 4]
```diff
@@ -69,3 +69,30 @@ export function getGeoDataIndex(buckets: VisualizationObject.IBucket[]): IGeoDat
 
     return geoDataIndex;
 }
+
+export function getGeoAttributeHeaderItems(
```
Comment (naibinh) [inline src/helpers/geoChart.ts]: Please consider this in my PR
https://github.com/gooddata/gooddata-react-components/pull/1315/files#diff-4e077217db022c533911725a328eb2feR158

File: src/components/core/GeoChart.tsx [hunk 5]
```diff
@@ -69,6 +83,31 @@ export class GeoChartInner extends BaseVisualization<IGeoChartInnerProps, {}> {
...
+        if (!execution) {
+            return;
+        }
+
+        const geoDataIndex: IGeoDataIndex = getGeoDataIndex(buckets);
+        if (
+            isDataOfReasonableSize(
+                execution.executionResult,
+                geoDataIndex,
+                limit || DEFAULT_DATA_POINTS_LIMIT,
```
Comment (naibinh) [inline src/components/core/GeoChart.tsx]: `limit` default prop should be `DEFAULT_DATA_POINTS_LIMIT `, right?

File: src/components/core/geoChart/GeoChartRenderer.tsx [hunk 6]
```diff
@@ -33,16 +34,26 @@ export default class GeoChartRenderer extends React.PureComponent<IGeoChartRende
     private chart: mapboxgl.Map;
     private chartRef: HTMLElement;
 
-    public componentDidMount() {
-        this.createMap();
+    public componentDidUpdate(prevProps: IGeoChartRendererProps) {
+        if (!isEqual(this.props.execution, prevProps.execution)) {
```
Comment (naibinh) [inline src/components/core/geoChart/GeoChartRenderer.tsx]: We should move this check to `shouldComponentUpdate` as above comment

File: src/components/core/GeoChart.tsx (hunk 3 above)
Comment (ptquang86) [reply src/components/core/GeoChart.tsx]: `validateData` function doesn’t stop the render flow. 
It only adds an `error` property into `BaseVisualization`.
`BaseVisualization` will use this `error` property to render the error screen

**Comment by ptquang86** [comment]
@naibinh @dlcd2203 Please help me review again

File: src/components/core/GeoChart.tsx (hunk 5 above)
Comment (ptquang86) [reply src/components/core/GeoChart.tsx]: updated

File: src/components/core/geoChart/GeoChartRenderer.tsx (hunk 6 above)
Comment (ptquang86) [reply src/components/core/geoChart/GeoChartRenderer.tsx]: fixed

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5686/ : SUCCESS in 16m 22s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1232/ : SUCCESS in 7s

File: src/helpers/geoChart.ts (hunk 4 above)
Comment (ptquang86) [reply src/helpers/geoChart.ts]: they are different. 1 used executionReponses, 1 used executionResults

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5690/ : FAILURE in 1m 50s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1238/ : SUCCESS in 6s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5692/ : SUCCESS in 5m 27s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1240/ : SUCCESS in 5s

**Comment by naibinh** [comment]
LGTM

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2911/ : SUCCESS in 16m 36s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5699/ : SUCCESS in 4m 48s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/1366/ : SUCCESS in 13s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/1892/ : FAILURE in 14m 26s

**Comment by ptquang86** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1324.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1250/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2336/ : SUCCESS in 16m 35s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2912/ : SUCCESS in 13m 41s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5700/ : SUCCESS in 4m 34s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/1367/ : SUCCESS in 15s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/1894/ : SUCCESS in 10m 27s

**Pull request merged**